Use presentation feedback

yol opened the pull request:
@FernetMenta this will be the most controversial PR yet, I think. 

I'll just let you have a look at it first and then we can discuss whether this is useful as-is.

pkerling committed:
Make private stuff in CWinSystemWayland private instead of protected or public

pkerling committed:
Add GetRefreshInHz() convenience function

pkerling committed:
Move SwapBuffers() call from derived classes to base

pkerling committed:
Add simple signal handler implementation

pkerling committed:
Add log category for timing information

This is useful for debugging AV sync etc. but should usually be disabled
because it generates a lot of information.

pkerling committed:
Log RenderManager frame timings

pkerling committed:
Bind and use wp_presentation for monitoring of frame timing

pkerling committed:
Implement and use CVideoSync implementation using wp_presentation

pkerling committed:
Reorganize WinSystemWayland members

Group them for clarity

pkerling committed:
Let windowing system determine display latency

instead of using hardcoded number of frames and multiplying with
refresh rate

pkerling committed:
Return display latency from wp_presentation

pkerling committed:
Fix EGL renderingApi parameter type

pkerling committed:
Remove clock offset from CVideoReferenceClock

The current calculation is weird and the DVD clock which is the only
user of this class has its own offset correction anyway.

pkerling committed:
Fix skin reload deadlocks in DisplaySettings once and for all

FernetMenta commented on xbmc/windowing/WinSystem.cpp:
```diff
@@ -254,6 +254,11 @@ std::string CWinSystemBase::GetClipboardText(void)
 
+float CWinSystemBase::GetDisplayLatency()
+{
+  return (NoOfBuffers() + 1) / g_graphicsContext.GetFPS();
```
windowing should not call into graphicsContext

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp:
```diff
@@ -1083,9 +1085,7 @@ void CRenderManager::PrepareNextRender()
 
-  // correct display latency
-  // internal buffers of driver, assume that driver lets us go one frame in advance
-  double totalLatency = DVD_SEC_TO_TIME(m_displayLatency) - DVD_MSEC_TO_TIME(m_videoDelay) + 2* frametime;
```
2*frametime is required, do not remove

FernetMenta replied:
I see, it is in windowing now
but windowing dows not know fps, see my other comment
maybe we should route this via gfx context instead of windowing

yol replied:
Yes it's quite funny that windowing which sets the fps in the first place does not know about it... Well, there *is* `m_fps` on `CWinSystemBase` though. Do you know whether it is reliable?

FernetMenta replied:
I don't see m_fps in CWinSystemBase

yol replied:
`m_fRefreshRate`, my bad

FernetMenta replied:
not sure what kind of a relict that is: https://github.com/xbmc/xbmc/search?utf8=%E2%9C%93&q=m_fRefreshRate&type=
don't use it

yol replied:
Mh what do we want to do here then. It seems CGraphicContext is the only one to (more or less...) reliably remember the FPS which is needed for the default calculation. But I can't put the Wayland stuff in there obviously. How about this: I remove the default calculation with the buffer count from `CWinSystemBase` and have it return some kind of invalid value by default. `CRenderManager` can then do the default calculation itself as fallback and ask `CGraphicContext` for the FPS.

FernetMenta replied:
RM can call GraphicsContext, then gfx context can call windowing for the remaining info.

off topic:
I was thinking more about usage of g_windowing and I am inclined to drop direct use of this global in general. 
For platform specific functions implementations could call something like CGraphicsContext::GetWindowingInterface and do a dynamic cast in order to get a specific interface.

yol replied:
Hm if we move this into GraphicsContext anyway, would you agree that it makes sense to move the `g_advancedSettings.GetDisplayLatency()` call which applies the overrides from advancedsettings to GraphicsContext, too?

> I was thinking more about usage of g_windowing and I am inclined to drop direct use of this global in general.

Yeah most places shouldn't need to use it directly. You have to consider that g_Windowing is *also* the RenderSystem instance which is used as such all over the code though. Maybe that should be separated anyway.

FernetMenta replied:
> would you agree that it makes sense to move the g_advancedSettings.GetDisplayLatency()

I don't think GraphicsContext should bother with this hack. why do you need it in GraphicsContext?

yol replied:
It's not that it's directly needed in GC, but also the uncorrected value is rarely needed and it's easy to forget about the correction. It'll probably make no difference since only RM uses it atm anyway. I'll just leave it there then.

FernetMenta commented on xbmc/cores/VideoPlayer/VideoReferenceClock.cpp:
```diff
@@ -206,7 +202,7 @@ int64_t CVideoReferenceClock::GetTime(bool interpolated /* = true*/)
   }
   else
   {
-    return CurrentHostCounter() + m_ClockOffset;
```
nice, I always wondered what this might be good for myself :)

yol replied:
Heh ok... just removed it and everything worked fine, so I assumed it's some leftover code. Also it didn't really make any sense.

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp:
```diff
@@ -1083,9 +1085,7 @@ void CRenderManager::PrepareNextRender()
-  // correct display latency
-  // internal buffers of driver, assume that driver lets us go one frame in advance
-  double totalLatency = DVD_SEC_TO_TIME(m_displayLatency) - DVD_MSEC_TO_TIME(m_videoDelay) + 2* frametime;
+  double totalLatency = DVD_SEC_TO_TIME(TotalLatency());
```
TotalLatency is measured from PresentRender. In order to be correct we also need the time from PresentRender untit here.

yol replied:
Not sure, shouldn't we include the time from here to PresentRender instead? If we take TotalLatency to be the time from selecting a frame to it being visible on screen. I admit I find it quite confusing though.

FernetMenta replied:
> shouldn't we include the time from here to PresentRender instead?

you mean having it as a separate value and adding it to TotalLatency? That would be ok

yol replied:
Yes separate value would be better, but the question was: What time do we need exactly?

* The time from `PresentRender` to `RenderManager::PrepareNextRender`, or
* The time from `RenderManager::PrepareNextRender` to `PresentRender`?

I think the second one would be correct.

FernetMenta replied:
iirc for the calculation we need the first: https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp#L1108

yol replied:
Ah `SetVsyncAdjust`, another mysterious influence on the timing :-) But doesn't that only kick in when the video fps matches the refresh rate almost exactly iirc? I must also admit that I'm not too sure what it exactly it is supposed to do. Why does it use `frametime / 2`? https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp#L1110

FernetMenta replied:
> Why does it use frametime / 2? 

RM has to decide whether to pick the next frame, skip it, or render the last frame again. frametime/2 moves the decision point to the middle of the frame. Then judgement is most reliable.

yol replied:
Okay, so why is the `frametime / 2` adjustment only done on the renderPts and not propagated to AV sync via `SetVsyncAdjust`? https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp#L1122

Also `frametime` is the output vsync interval and not the interval between video frames.

FernetMenta replied:
Don't mix up things. We deal with:
1) a/v sync
Presentation timestamp of a video frame relates to the time the frame becomes visible. In order to sync audio to this time we need the offset beween vertical sync signal and video pts.

2) queue video frames into GL.
When the render threads comes by we need to decide if the next frame in our render buffers is due. This is not completely trival because video drivers may queue frames internally. When playing some 50 fps content for example, the render thread can be blocked close to 40ms and we don't necessarily lose a frame. The next two cycles will be done in much less than frametime. RM needs to account for this elasticity. Moving the decision point to t(vsync) + frametime / 2 makes it most reliable.

yol replied:
I'm aware of that distinction in principle, but it's very easy to get confused here...

fwiw on Wayland with mesa and usual compositors you will probably lose at least one frame in your example of blocking for ~40 ms. Even if you render very fast afterwards and immediately submit the two rendered frames to the compositor, it will typically just discard the earlier buffer if it hasn't used it yet. Also, mesa synchronizes with the compositor redraw cycle by blocking in eglSwapBuffers when eglSwapInterval(1) is used, so you cannot render faster than vrefresh in that case.

Sorry for the sidetracking - let's get back to the main issue of correcting the latency so this moves forward :-) Should I take the time from `PresentRender` at function exit or enter, i.e. before or after `eglSwapBuffers` (makes a big difference)? And is this similar to https://github.com/FernetMenta/kodi-agile/blob/master/xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp#L1097 or something else altogether?

FernetMenta replied:
> so you cannot render faster than vrefresh in that case

but we get one frame ahead. assume we have not called swapBuffers for a while. now the first call to swapBuffers will not block because of double buffers.

> Should I take the time from PresentRender at function exit or enter, i.e. before or after eglSwapBuffers (makes a big difference)?

if swapBuffers blocks, it makes a difference.

> And is this similar to https://github.com/FernetMenta/kodi-agile/blob/master/xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp#L1097 or something else altogether?

this code gets the time since last vblank. not sure if I get your question

yol replied:
> but we get one frame ahead. assume we have not called swapBuffers for a while. now the first call to swapBuffers will not block because of double buffers.

Yes, eglSwapBuffers will not block on the call after the 40 ms block, since it already received the frame sync from the compositor at that point. But it will block again right at the next call after that to resync with the compositor even if you have rendered faster. I assume on X/intel you could catch up because there are more buffers on the client side?

> if swapBuffers blocks, it makes a difference.

Exactly, that's why I asked: Take the time before or after swapBuffers?

> this code gets the time since last vblank. not sure if I get your question

Question was if that is somehow related to the problem at hand. Guess the answer means "no" ;-)

FernetMenta replied:
> I assume on X/intel you could catch up because there are more buffers on the client side?

well, two buffers are enough. I would have thought by requesting a visual with double buffers we get two buffers. On X11 you can clearly see that we get a frame ahead and if we are not a frame ahead, there is no blocking.

> Exactly, that's why I asked: Take the time before or after swapBuffers?

We need the time from last vblank. If swapbuffers blocks, then after. On X11 it is not that trivial, because swapBuffers does not necessarily block (see above and X11 on my branch)

yol replied:
> I would have thought by requesting a visual with double buffers we get two buffers.

Mesa-egl on Wayland does not use a fixed number of buffers (don't know about X). Every time you call eglSwapBuffers, it submits the buffer that was currently being rendered into to the compositor. The compositor uses it to draw the screen. When screen redraw is done and the buffer will not be used anymore, the compositor tells mesa so that it can use the buffer for drawing again (if mesa would reuse the buffer before that it would lead to tearing). Mesa will allocate and cycle through as many buffers as needed such that drawing can happen (after eglSwapBuffers) on a buffer that is not in use by the compositor.

> We need the time from last vblank. If swapbuffers blocks, then after. On X11 it is not that trivial, because swapBuffers does not necessarily block (see above and X11 on my branch)

On Wayland, eglSwapBuffers also does not necessarily block (no blocking if we are late and missed the last frame hint), but also if it does block it does not follow vblank timing. swapBuffers is synchronized to vblank in the sense that if you render fast enough and have an output with 60 Hz, swapBuffers will block so that you render 60 fps on average. But the exact point in time swapBuffers returns is usually somewhere in the middle between two vblanks, when the compositor decides that it would be a good time for drawing.

FernetMenta replied:
According to your comment in CWinSystemWayland::PrepareFramePresentation we should be able to calculate vblank time. A frame becomes visible short after vblank and m_latencyMovingAverage is average time from swapBuffers to visible (at a vblank). right?

At the time of swapBuffers ptsOnScreen should be clock - m_latencyMovingAverage. Later in PrepareNextRender the clock has advanced t(PrepNextRender) - t(swapBuffers). I would subtract this time from clock to get ptsOnScreen.

yol replied:
> we should be able to calculate vblank time

wp_presentation tells us when the vblank happens for each frame we commit to the compositor with swapBuffers

> A frame becomes visible short after vblank and m_latencyMovingAverage is average time from swapBuffers to visible (at a vblank). right?

Yes, more concretely latencyMovingAverage is the average time from before eglSwapBuffers call (i.e. when PresentRender is called) to visible. So it will include the time eglSwapBuffers() blocks.

> At the time of swapBuffers ptsOnScreen should be clock - m_latencyMovingAverage. Later in PrepareNextRender the clock has advanced t(PrepNextRender) - t(swapBuffers). I would subtract this time from clock to get ptsOnScreen.

OK, that sounds like I need the time from after eglSwapBuffers(), right? I'll try that.

FernetMenta replied:
ok, let's got for this. we still can change it :)

FernetMenta commented on xbmc/commons/ilog.h:
```diff
@@ -53,6 +53,7 @@
 #define LOGCEC        (1 << (LOGMASKBIT + 9))
 #define LOGVIDEO      (1 << (LOGMASKBIT + 10))
 #define LOGWEBSERVER  (1 << (LOGMASKBIT + 11))
+#define LOGTIMING     (1 << (LOGMASKBIT + 12))
```
not sure if such a generic specifier is a good idea. it will get used by different components and you'll end up seeing a lot of python log spam too.

yol replied:
I specifically mean audio and video decode/rendering timing, so maybe `LOGAVTIMING` is better? Or is it still too generic?

FernetMenta replied:
let's try this :)

yol replied:
done

FernetMenta commented on xbmc/utils/Signals.h:
```diff
@@ -0,0 +1,201 @@
+/*
```
I think for now it is better to keep those things private to wayland. Kodi already has too many different approaches how to deal with signals/messages. 
IMO we should move to an actor model with asynchronous message passing.

yol replied:
I tried to look around how other parts of the Kodi code do it, but none of those quite 'fit' - might also be because a large part of the code base predates C++11. I agree that it would be better to find a set of common models that the code uses consistently, if only for new code in the beginning. I'll just move this to the Wayland folder for now. If you have concrete ideas on how the "future" should look like we can try them out here though :-)

FernetMenta replied:
> If you have concrete ideas on how the "future" should look like we can try them out here though :-)

Have a look at ActiveAE. The concept is based on UML ports and protocols. The state machine makes it easy to separate aspects of behaviour.

yol replied:
Nice, and the actor stuff is already in util. It's also very optimized, I see. But I'm not sure it would solve the problem I have here and why I made the custom signal stuff in the first place:

Windowing is asked for a new video sync implementation, but it does not own it afterwards - the caller does. And from the interface it is not guaranteed that only one such video sync objects exists - so theoretically, there could be two or more video sync objects active at the same time. They all need to get the presented event, so the signal mechanism must support multiple subscribers.

Otherwise, I'd have just used a plain interface or single std::function.

FernetMenta replied:
I am fine with this new approach when kept private in the first round. Later we should start some consolidation and agree when to use what pattern. (still hoping that this will happen sometime :)

FernetMenta replied:
> They all need to get the presented event, so the signal mechanism must support multiple subscribers.

for the record. ActorProtocal supports connecting ports by one to many :)

yol replied:
Moved for now as suggested. ActorProtocol can really do anything can't it :-) fwiw the signal/slot approach here is similar to what most toolkits like Gtk+ and Qt do and basically a (very) bare-bones version of Boost.Signals2.

pkerling commented:
I think that this calculation is wrong: https://github.com/pkerling/xbmc/blob/6d806f359f9a35a29107b2cfb211b3793bce99c5/xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp#L1203

From what I can tell, the intention here is to enable the VSyncAdjust only when screen refresh rate and adjusted video FPS are (almost or exactly) an integer multiple of each other. But this is not what the code checks in some cases.

Take for example: Screen refresh rate is 60 Hz. Video FPS is 30.001 (because of limited precision). Then the code calculates 60 modulo 30.001, which is 29.999. So VSyncAdjust will be disabled although 60 and 30.001 are a very good match. For screen refresh rate 60 Hz and video fps 29.999 on the other hand it would be enabled since 60 modulo 29.999 is 0.002, which I think is strange.

Suggestion:

```
diff = g_graphicsContext.GetFPS() / fps;
if (diff < 1.0)
  diff = 1.0 / diff;
// Calculate relative distance from nearest integer proportion
diff = std::abs(std::round(diff) - diff);
```

This would yield `diff = 0.00006...` for the example above. The check would then be `diff < 0.0005` instead of `diff < 0.01` since it is now relative instead of absolute.

FernetMenta commented:
Ok, go ahead with your suggestion.
btw: I have not seen values for video fps like in your eaxamples in real world. Video fps follows discrete values and as a result it cannot suffer from precision errors. maybe variable frame rafe material but this is out of scope for this use case anyway.

pkerling commented:
Yes the problem is not the video itself, but if you adjust it with the DVDClock speed, it will become imprecise somewhere (not sure where exactly, probably some kind of rounding or float/double issue). I have seen this in the example I wrote (30 fps video), where the DVDClock speed is calculated to be 1.0000000000000002, leading to 30.0000000000000007 fps - which is enough to trigger the problem.

pkerling committed:
Fix RenderManager VSyncAdjust enable condition

Previous code was wrong because it fails on slightly imprecise values.
For example fmod(60,29.999) is 0.002, which is OK.
But fmod(60,30.001) yields is 29.999, which will get rejected as being
above 0.01. There is no reason why 30.001 should be rejected while
29.999 is accepted.

pkerling committed:
Fix xdg_shell ping handler not deregistered on destruction

FernetMenta commented:
> Yes the problem is not the video itself, but if you adjust it with the DVDClock speed

we do not adjust video fps.

pkerling commented:
When you adjust the video clock speed, naturally you will get slightly different fps. Talking about this: https://github.com/pkerling/xbmc/blob/6d806f359f9a35a29107b2cfb211b3793bce99c5/xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp#L1199

FernetMenta commented:
right, sorry. I almost forgot about this :)

pkerling committed:
Fix crash on shutdown

pkerling committed:
Subtract time from the start of frame painting in PrepareNextRender for renderPts

Signed-off-by: Philipp Kerling <pkerling@casix.org>

pkerling commented:
Last outstanding issue fixed, rebased, squashed
@FernetMenta please check :)

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp:
```diff
@@ -926,11 +926,13 @@ void CRenderManager::UpdateDisplayLatency()
-
-  int buffers = g_Windowing.NoOfBuffers();
-  m_displayLatency += (buffers - 1) / fps;
+  m_displayLatency = g_advancedSettings.GetDisplayLatency(refresh);
```
instead of changing the meaning of m_displaylatendy and introducing a new method TotalLatency() (just for debug purposes it stresses gfx context twice), you should update m_displayLatency in PrepareNextRender().  then the debug output can use m_displayLatency which is better anyway because you see the value that is actually used and not somthing updated at a different time

yol replied:
The latency overall was totally static before which is no longer possible. If we reuse m_displayLatency for the dynamically updated latency, we would have to call `CAdvancedSettings::GetDisplayLatency` on every frame - at the moment it is only called when resolution changes. But I agree that it would be better to show the actually used value in the UI. So what do you think: Call `CAdvancedSettings::GetDisplayLatency` every frame anyway or introduce an extra variable for the dynamic part?

FernetMenta replied:
the latter seems to be a good idea but maybe we should introduce a new variable for the advanced setting because it's not really displaylatency but some tweak value nodoy has figured out so far why it's needed on some systems and not on others.

yol replied:
done

FernetMenta commented on xbmc/windowing/WinSystem.h:
```diff
@@ -85,6 +85,21 @@ class CWinSystemBase
+  virtual float GetDisplayLatency() { return -1.0f; }
+  /**
+   * Get time that has elapsed since the start of the rendering the current frame
+   * in seconds
```
ms are better suited for this purpose

yol replied:
AdvancedSettings::GetDisplayLatency is also in seconds - shouldn't it be consistent?

FernetMenta replied:
there are a lot of different time units used in Kodi. sure consistency is desired but not in seconds. IMO ms fits best for player.

yol replied:
Should `GetDisplayLatency` be changed to ms too?

FernetMenta replied:
Do you mean windowing or advancedSettings? if you change the latter, please change the name of the method too.

yol replied:
Windowing, mainly. But we can as well change all. With renaming, do you mean `AdvancedSettings::GetDisplayLatency` -> `AdvancedSettings::GetLatencyTweak` or something else?

yol replied:
Wait, you didn't even see that commit yet :-) Sorry for the confusion. Just tell me which functions I should change to milliseconds and how I should rename them.

FernetMenta replied:
> `AdvancedSettings::GetDisplayLatency` -> `AdvancedSettings::GetLatencyTweak`

that would be great :)

all units should be in ms

yol replied:
Renamed and all values participating in m_displayLatency changed to ms unit

FernetMenta commented on xbmc/windowing/WinSystem.h:
```diff
@@ -85,6 +85,21 @@ class CWinSystemBase
+   * Get time that has elapsed since the start of the rendering the current frame
+   * in seconds
+   */
+  virtual float GetTimeSinceFrameStart() { return 0.0; }
```
we need a more genering name. on other systems this is time since last vblank

yol replied:
Tough to find the common ground here. `GetTimeSinceBufferSwap`? `GetTimeSincePresentRender`? Or even more generic: `GetFrameLatencyAdjustment`?

FernetMenta replied:
I like GetFrameLatencyAdjustment most

yol replied:
done

FernetMenta commented on xbmc/windowing/WinSystem.h:
```diff
@@ -95,6 +95,15 @@ class CWinSystemBase
   virtual float GetDisplayLatency() { return -1.0f; }
+  /**
+   * Get time that should be subtracted from the display latency for this frame
+   * in milliseconds
```
great

pkerling committed:
Rename AdvancedSettings::GetDisplayLatency -> GetLatencyTweak, change unit

New unit is milliseconds (instead of seconds) since most of the player
code uses it.

Renaming because the time is not related to the "real" display latency
that can somehow be measured, but rather an unspecific tweak.

Signed-off-by: Philipp Kerling <pkerling@casix.org>

pkerling commented:
And merged, big thanks! Windowed mode is almost ready for PR, that will be another big one ;-)
